PyTorch nn Layers entry (#5586)
Codecademy/docs: content/pytorch/concepts/nn/terms/layers/layers.md

---
Title: 'layers'

> Description: 'Layers and their role in PyTorch neural networks.' 

Reviewer: The description should tell what layers are and it should be repeated in the introductory paragraph of the entry as per the [style guide](https://github.com/Codecademy/docs/blob/main/documentation/style-guide.md). 

We can define it as following -

```suggestion
Description: 'Layers are the fundamental building blocks used to construct neural networks.' 
```
or 

```suggestion
Description: 'Layers are modules that perform operations on input data to build neural networks.'
```

Please update the same in the first paragraph based on your choice.

Author: Updated description with option 2 to reflect new opening paragraph.

Subjects:
  - 'AI'
  - 'Machine Learning'
Tags:
  - 'Machine Learning'
  - 'Neural Networks'
  - 'AI'
  - 'Deep Learning'
  - 'Models'
CatalogContent:
  - 'learn-intro-to-py-torch-and-neural-networks'

>   - 'paths/machine-learning'

Author: Thank you for the resource.  I will get this updated.

---

> **Layers** are higher-level constructs built from **Tensors**, used to define the architecture of neural networks in PyTorch. Neural networks are created by stacking layers in specific patterns to transform input data into outputs, such as image classifications or predictions. PyTorch offers various layers, including **Linear**, **Convolutional**, **Recurrent**, **Pooling**, and **Transformer layers**, each optimized for tasks like image recognition or natural language processing. By stacking multiple layers, neural networks can learn complex patterns, making them effective for a variety of AI applications.

Author: Updated with suggestion 2.

> 

Reviewer: We should also skip the names of different layers for simplicity here as the reader is still trying to understand what are layers as they are reading and this is a step ahead.

## Syntax

> By using PyTorch's sequential() method to build a neural network, we can specify **layers** and activation functions in sequence from input to output as show below:

Reviewer: When referring to properties and methods, we should mention the exact function name along with `.` and `()` to help the user relate the text to the actual syntax.

Also, bold should only be used when the term is introduced for the first time. 

The ending `n` is missing in `shown`.

```suggestion
By using PyTorch's `.Sequential()` method to build a neural network, we can specify layers and activation functions in sequence from input to output as shown below:
```

Author: Updated to suggested change.

```pseudo
import torch
import torch.nn as nn

# Define the model for the neural network

> model = nn.Sequential(
>     # Define layers and activation functions here
>     nn.Linear(in_features, out_features)
>     nn.ReLU()
>     nn.Linear(in_features, out_features)
> )

Reviewer: In `nn.Sequential()` method, each layer should be separated by a comma. 

Here’s the corrected version:

```suggestion
model = nn.Sequential(
    # Define layers and activation functions here
    nn.Linear(in_features, out_features),
    nn.ReLU(),
    nn.Linear(in_features, out_features)
)
```

The commas are essential in Python to separate items within argument lists like `nn.Sequential()`.

```

## Example

> Below is an example of a basic Multi-**Layer** sequential neural network where data flows down through the specified sequence of layers and activation functions:

Reviewer: The term 'Layer' should be highlighted in bold only when it is used for the first time. 

The term "Multi-Layer" should be written as "multi-layer" (lowercase and hyphenated) as it is not a part of a title.

```suggestion
Below is an example of a basic multi-layer sequential neural network where data flows down through the specified sequence of layers and activation functions:
```

Reviewer: The phrase "data flows down" could be rephrased to use active voice for clarity. For example, "where the data is processed in a top-down manner" could make the action clearer.

Here is the complete sentence for reference - 

```suggestion
Below is an example of a basic multi-layer neural network defined using PyTorch's `nn.Sequential()` method, where the data is processed through a specified sequence of layers and activation functions:
```

> ```py
> import torch
> import torch.nn as nn
> 
> model = nn.Sequential(
>     nn.Linear(56,128),
>     nn.ReLU(),
>     nn.Linear(128,26),
>     nn.ReLU(),
>     nn.Linear(26,1)
> )
> ```

Reviewer: I would suggest adding relevant code comments to tell the user what is happening in different steps.

```suggestion
```py
import torch
import torch.nn as nn

model = nn.Sequential(
    # First layer: Linear transformation from 56 input features to 128 output features
    nn.Linear(56, 128),  # This layer learns to transform input data
    nn.ReLU(),           # Activation function to introduce non-linearity
    
    # Second layer: Linear transformation from 128 input features to 26 output features
    nn.Linear(128, 26),  # This layer continues to transform the data
    nn.ReLU(),           # Another activation function for non-linearity
    
    # Third layer: Linear transformation from 26 input features to 1 output feature
    nn.Linear(26, 1)     # Final output layer that produces the model's prediction
)
```
```

---
Title: 'Layers'
Description: 'Layers are modules that perform operations on input data to build neural networks.' 
Subjects:
  - 'AI'
  - 'Machine Learning'
Tags:

>   - 'Machine Learning'
>   - 'Neural Networks'
>   - 'AI'
>   - 'Deep Learning'
>   - 'Models'

Reviewer 2: The tags should be in alphabetical order.
```suggestion
  - 'AI'
  - 'Deep Learning'
  - 'Machine Learning'
  - 'Models'
  - 'Neural Networks'
```

CatalogContent:
  - 'learn-intro-to-py-torch-and-neural-networks'
  - 'py-torch-for-classification'
  - 'paths/machine-learning'
---

> Layers are modules that perform operations on input data to build neural networks.

Reviewer 2: ```suggestion
**Layers** are modules that perform operations on input data to build neural networks.
```
